@author: Carson Hanel

STAT 489: Principles of Data Science and Statistics

Professor: Alan Dabney

Homework: Initial Python

Task 1: Create four Python variables with different data types:
----

In [59]:
import numpy as np
import math
import random

names     = ["Bob", "Jill", "George", "Anne"] # String
salaries  = [89900,  48000, 128000  ,  56000] # Integer
raise_p   = [2.3  ,  1.5  , 3.0     ,  1.2  ] # Float
bonus     = [2500 ,  1200 , 5200    ,  1400 ] # Integer
retire    = [True ,  False, True    ,  True ] # Boolean
salary_p  = [[]   ,  []   , []      ,  []   ] # Will hold lists of projected 5 year salaries

name_dict = {"Bob"    : 0,
             "Jill"   : 1,
             "George" : 2,
             "Anne"   : 3,
             "Eric"   : 4,
             "Mandy"  : 5};

Task 2: Create a Python list containing 5 numeric variables:
----

Note: I know this is extra. I was just having fun. I created a function to generate everyone's 5 year salary outlook.

In [5]:
def calculate_half_decade(salary, raise_p):
    salaries    = np.zeros(5)
    salaries[0] = salary * ((100 + raise_p)/100)
    for i in range(1, 5):
        salaries[i] = salaries[i - 1] * ((100 + raise_p)/100)
    return salaries

def print_salary(salaries):
    salary  = "["
    for element in salaries:
        salary += " " + str(element) + ","
    salary += "]"
    print(salary)
    
for i in range(len(salaries)):
    salary_p[i] = calculate_half_decade(salaries[i], raise_p[i])

for i in range(len(names)):
    print(str(names[i]) + "'s five year salary outlook: ")
    print_salary(salary_p[i])
    print("\n")

Bob's five year salary outlook: 
[ 91967.7, 94082.9571, 96246.8651133, 98460.5430109, 100725.1355,]


Jill's five year salary outlook: 
[ 48720.0, 49450.8, 50192.562, 50945.45043, 51709.6321864,]


George's five year salary outlook: 
[ 131840.0, 135795.2, 139869.056, 144065.12768, 148387.08151,]


Anne's five year salary outlook: 
[ 56672.0, 57352.064, 58040.288768, 58736.7722332, 59441.6135,]




Task 3: Perform an arithmetic operation on a subset of elements in this list:
----

Note: So, what I'll do now is assume that every other year, everyone gets a bonus!

In [6]:
def insert_raise(salaries, amount):
    for i in range(len(salaries)):
        if (i-1)%2 == 0:
            salaries[i] += amount
    return salaries

for i in range(4):
    salary_p[i] = insert_raise(salary_p[i], bonus[i])

for i in range(len(names)):
    print(str(names[i]) + "'s five year salary outlook with added bonuses: ")
    print_salary(salary_p[i])
    print("\n")

Bob's five year salary outlook with added bonuses: 
[ 91967.7, 96582.9571, 96246.8651133, 100960.543011, 100725.1355,]


Jill's five year salary outlook with added bonuses: 
[ 48720.0, 50650.8, 50192.562, 52145.45043, 51709.6321864,]


George's five year salary outlook with added bonuses: 
[ 131840.0, 140995.2, 139869.056, 149265.12768, 148387.08151,]


Anne's five year salary outlook with added bonuses: 
[ 56672.0, 58752.064, 58040.288768, 60136.7722332, 59441.6135,]




Cool! So now you can see, on "even" years (if counting from year of employment 1...) everyone got bonuses!

Task 4: Create a Python list containing 3 string values
----

Task 5: Append 2 more strings to the end of this python list
----

To address this, I've already got the list of names of people. In order to satisfy these requirements, I'll add two more people.

Just for continuity, now that they're there, I'll also calculate their 5-year. Why not?

In [7]:
names    += ["Eric", "Mandy" ] # String
salaries += [15500 ,  98000  ] # Integer
raise_p  += [1     ,  4.5    ] # Float
bonus    += [200   ,  3200   ] # Integer
retire   += [False ,  True   ] # Boolean
salary_p += [[]    ,  []     ] # Will hold lists of projected 5 year salaries

print("We've got some new employees! Please greet: " + " and ".join(names[4:6]) + "!")

for i in range(4, 6):
    salary_p[i] = calculate_half_decade(salaries[i], raise_p[i])
    salary_p[i] = insert_raise(salary_p[i], bonus[i])

for i in range(len(names)):
    print(str(names[i]) + "'s five year salary outlook with added bonuses (Updated): ")
    print_salary(salary_p[i])
    print("\n")

We've got some new employees! Please greet: Eric and Mandy!
Bob's five year salary outlook with added bonuses (Updated): 
[ 91967.7, 96582.9571, 96246.8651133, 100960.543011, 100725.1355,]


Jill's five year salary outlook with added bonuses (Updated): 
[ 48720.0, 50650.8, 50192.562, 52145.45043, 51709.6321864,]


George's five year salary outlook with added bonuses (Updated): 
[ 131840.0, 140995.2, 139869.056, 149265.12768, 148387.08151,]


Anne's five year salary outlook with added bonuses (Updated): 
[ 56672.0, 58752.064, 58040.288768, 60136.7722332, 59441.6135,]


Eric's five year salary outlook with added bonuses (Updated): 
[ 15500.0, 15700.0, 15500.0, 15700.0, 15500.0,]


Mandy's five year salary outlook with added bonuses (Updated): 
[ 102410.0, 110218.45, 111834.28025, 120066.822861, 122125.82989,]




Task 6: Combine your two Python lists into a single two-dimensional list.
----

Alright, so you decide you want a matrix of just salary outlook data.

In [8]:
outlook_prediction = []
for prediction in salary_p:
    outlook_prediction.append(prediction)
for person_data in outlook_prediction:
    print(person_data)

[  91967.7          96582.9571       96246.8651133   100960.54301091
  100725.13550016]
[ 48720.          50650.8         50192.562       52145.45043
  51709.63218645]
[ 131840.         140995.2        139869.056      149265.12768
  148387.0815104]
[ 56672.          58752.064       58040.288768    60136.77223322
  59441.61350001]
[ 15500.  15700.  15500.  15700.  15500.]
[ 102410.          110218.45        111834.28025     120066.82286125
  122125.82989001]


Now, to augment this new data, the boss decides he wants to add a yearly five thousand dollar buffer on the top of those employees that are given a retirement by the company in order to represent company contributions to 401k and retirement options.

To do this, I'll write a function.
Task 7: Use any five Python functions:
----
    - 1) calculate_half_decade()
    - 2) print_salary()
    - 3) insert_raise()
    - 4) retirement_cost()
    - 5) sum()
    
Task 8: Use any five Python methods:
----
    - 1) np.zeros()
    - 2) variable.append()
    - 3) " ".join(variable)
    - 4) random.choice(names)
    - 5) random.random()
    

In [9]:
def add_retirement(salaries, status):
    for i in range(len(salaries)):
        if status[i]:
            for years in salaries[i]:
                years += 5000
    return salaries

salary_p = add_retirement(salary_p, retire)

for i in range(len(names)):
    print(str(names[i]) + "'s five year outlook with bonuses and retirement: ")
    print_salary(salary_p[i])
    print("\n")

Bob's five year outlook with bonuses and retirement: 
[ 91967.7, 96582.9571, 96246.8651133, 100960.543011, 100725.1355,]


Jill's five year outlook with bonuses and retirement: 
[ 48720.0, 50650.8, 50192.562, 52145.45043, 51709.6321864,]


George's five year outlook with bonuses and retirement: 
[ 131840.0, 140995.2, 139869.056, 149265.12768, 148387.08151,]


Anne's five year outlook with bonuses and retirement: 
[ 56672.0, 58752.064, 58040.288768, 60136.7722332, 59441.6135,]


Eric's five year outlook with bonuses and retirement: 
[ 15500.0, 15700.0, 15500.0, 15700.0, 15500.0,]


Mandy's five year outlook with bonuses and retirement: 
[ 102410.0, 110218.45, 111834.28025, 120066.822861, 122125.82989,]




Just to supply the boss with some analytics, you decide to offer the average yearly cost for each employee, and what percentile greater than their base salary their benefits are.

In [10]:
def calculate_yearly_cost(salaries):
    return [sum(salary) / len(salary) for salary in salaries]

def calculate_benefit(salary_init, salaries):
    results = np.zeros(len(salary_init))
    for i in range(len(salary_init)):
        base_pay = salary_init[i] * len(salaries[i])
        benefits = 0
        for year in salaries[i]:
            benefits += (year - salary_init[i])
        results[i] = float(benefits / base_pay)
        
    return results

yearly_cost = np.zeros(len(names))
benefits    = np.zeros(len(names))

yearly_cost = calculate_yearly_cost(salary_p)
benefits    = calculate_benefit(salaries, salary_p)

print("Average Yearly Cost: ")
print(yearly_cost)
print("\nAverage Employee Bonus: ")
print(benefits)


Average Yearly Cost: 
[97296.640144872508, 50683.688923289985, 142071.29303808004, 58608.547700246119, 15580.0, 113331.07660025125]

Average Employee Bonus: 
[ 0.08227631  0.05591019  0.10993198  0.04658121  0.00516129  0.15643956]


Now, the metric for "Average Employee Bonus" doesn't tell us much.

In order to get an accurate gauge of who most benefits on a scale of 0-100, we must apply a feature scaling algorithm.

In [11]:
def mass_feature_scale(data):
    high = float(max(data))
    low  = float(min(data))
    
    for i in range(len(data)):
        data[i] = float(((data[i] - low) / (high - low)) * 100.)
    return data

scaled_benefits = mass_feature_scale(benefits)
print("Scaled Employee Bonus: ")
print(scaled_benefits)

Scaled Employee Bonus: 
[  50.97560933   33.54671933   69.2569321    27.37995317    0.          100.        ]


This still doesn't tell us much, as it is guaranteed that the outcome will contain a 0 and 100 value as it is scaling dependent on the maximum and minimum elements of the set. Really, this would be more useful in large companies trying to construct bucket ranges for their employees to see where they lay on different scaled company-driven percentiles.

Finally, the boss decides it's time to give back. 

In order to do this, he's requested you randomly select someone to recieve a one-time bonus between 2,500 and 5,000 dollars. Rather than do it manually, you decide it'd be a bit of fun to let the random number generator decide.

As a final note, your boss adds that he wants their index to be less than 50; he wants to help where it's needed.

In [34]:
def generate_winner(names, index, threshold):
    winner = random.choice(names)
    winner_index = index[name_dict[str(winner)]]
    while winner_index > threshold:
        winner = random.choice(names)
        winner_index = index[name_dict[str(winner)]]
    prize  = 2500 + random.random() * 2500
    return winner, prize

winner, prize = generate_winner(names, scaled_benefits, 50.)
print("And the winner is: " + str(winner) + "!")
print("They won: %.2f dollars!\n" % prize)

# Float formatting
print(str(winner) + " has a bonus index of          : %.2f" % scaled_benefits[name_dict[winner]])
print(str(winner) + " has an average yearly cost of : %.2f" % yearly_cost[name_dict[winner]])

# String casting
print(str(winner) + " has a raise percentile of     : " + str(raise_p[name_dict[str(winner)]]))
print(str(winner) + " gets a yearly bonus of        : " + str(bonus[name_dict[str(winner)]]))

And the winner is: Jill!
They won: 3561.10 dollars!

Jill has a bonus index of          : 33.55
Jill has an average yearly cost of : 50683.69
Jill has a raise percentile of     : 1.5
Jill gets a yearly bonus of        : 1200


Task 9: Create a two-dimensional NumPy array and print a subset of its rows and columns
----

So the boss sees that Jill has won 3561.10 and is feeling extra generous. He'd like to give the remaining 1438.9 to the person with the lowest bonus compensation percentile to try and spread the wealth. The boss requests that you locate that person and bring him their file.

Note: The subset that's being printed is the row containing the lowest bonus earner, and all columns within the row being printed.

In [57]:
def find_min_location(percentiles):
    min_i = 0
    for i in range(len(percentiles)):
        if percentiles[min_i] > percentiles[i]:
            min_i = i
    return min_i

def print_user(ID, names, salaries, salary_p, bonus, retire, raise_p, benefits):
    five_year  = "["
    for year in salary_p[ID]:
        five_year += " %.2f," % year
    five_year += "]"
    print("Employee Name      : " + str(names[ID]))
    print("Employee Pay       : %.2f" % salaries[ID] )
    print("Employee Five-Year : " + five_year)
    print("Employee Bonus     : %.2f" % bonus[ID])
    print("Employee Raise Pct : %.2f" % raise_p[ID])
    print("Employee Benefit   : %.2f / 100" % benefits[ID])
    print("Employee Retirement: " + str(retire[ID]))

lowest_compensated = find_min_location(scaled_benefits)
print_user(lowest_compensated, names, salaries, salary_p, bonus, retire, raise_p, scaled_benefits)
    

Employee Name      : Eric
Employee Pay       : 15500.00
Employee Five-Year : [ 15500.00, 15700.00, 15500.00, 15700.00, 15500.00,]
Employee Bonus     : 200.00
Employee Raise Pct : 1.00
Employee Benefit   : 0.00 / 100
Employee Retirement: False


Task 10: Calculate the mean, median, and standard deviation of the two variables in your NumPy array.
----

Finally, you end the day by printing some basic statistics about your employee dataset:

In [68]:
print("Mean of Salaries    : %.2f" % np.mean(salaries))
print("Median of Salaries  : %.2f" % np.median(salaries))
print("Standard Deviation  : %.2f\n" % np.std(salaries))

print("Mean of Bonus       : %.2f" % np.mean(bonus))
print("Median of Bonus     : %.2f" % np.median(bonus))
print("Standard Deviation  : %.2f\n" % np.std(bonus))

print("Mean of Raise Pct   : %.2f" % np.mean(raise_p))
print("Median of Raise Pct : %.2f" % np.median(raise_p))
print("Standard Deviation  : %.2f\n" % np.std(raise_p))

Mean of Salaries    : 80475.00
Median of Salaries  : 72950.00
Standard Deviation  : 31627.15

Mean of Bonus       : 2575.00
Median of Bonus     : 1950.00
Standard Deviation  : 1594.33

Mean of Raise Pct   : 2.00
Median of Raise Pct : 1.90
Standard Deviation  : 0.70

